### PR TITLE
PL-129868 structured haproxy config

### PR DESCRIPTION
**Pull request by elikoga**
@flyingcircusio/release-managers

## Release process

Impact:

- Haproxy service will be reloaded/restarted

Changelog:

- Adds nix options to configure haproxy under the key `flyingcircus.services.haproxy`
- Migrate internal services to haproxy configured with nix

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - Haproxy config can be read from nix store - has to be unconfidential
- [x] Security requirements tested? (EVIDENCE)
  - Haproxy can be configured through nix options
  - System still builds
  - haproxy.nix test passes

### CODE REVIEW

File: nixos/roles/graylog.nix [hunk 1]
```diff
@@ -249,92 +249,88 @@ in
...
-            option httpchk GET /
-            timeout server 121s    # should be equal to client timeout
-        ${backendConfig (name: ip:
-            "server ${name}  ${ip}:${toString glAPIPort} check fall 1 rise 2 inter 10s maxconn 20")}
-
-        backend stats
-            stats uri /
-            stats refresh 5s
-      '');
+      flyingcircus.services.haproxy = lib.mkOverride 90 {
```
Comment (dpausp) [inline nixos/roles/graylog.nix]: the override is not needed here, I think

File: nixos/roles/graylog.nix [hunk 2]
```diff
@@ -249,92 +249,88 @@ in
...
+        enable = true;
+        listens = {};
```
Comment (dpausp) [inline nixos/roles/graylog.nix]: this should not be neccessary

File: nixos/roles/graylog.nix [hunk 3]
```diff
@@ -249,92 +249,88 @@ in
...
+              client = "10s";
+            };
+            default_backend = "gelf_tcp";
+          };
+          beats-tcp-in = {
+            binds = map (addr: "${addr}:${toString beatsTCPHAPort}") (fclib.listenAddresses "ethsrv");
+            mode = "tcp";
+            options = [ "tcplog" ];
+            timeout = {
+              client = "121s"; # should be equal to client timeout
```
Comment (dpausp) [inline nixos/roles/graylog.nix]: should be "should be equal to server timeout?" We could reflect that fact in the structured config by setting the common value via `let beatsInputTimeout = "121s";`

File: nixos/services/haproxy/README.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+backends.<name>.options
+backends.<name>.default_backend
+backends.<name>.balance
+backends.<name>.servers
+backends.<name>.extraConfig
+extraConfig
+```
+
+Example configuration:
+
```
Comment (dpausp) [inline nixos/services/haproxy/README.md:133]: there should be an additional example configuration based on the current plain example config that we generate. Kubernetes as an example is a bit too special (but can remain here).

File: nixos/services/haproxy/README.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
+These modules allow you to generate a haproxy configuration from structured nix values.
+
+You have several avaliable options under `flyingcircus.services.haproxy`:
```
Comment (dpausp) [inline nixos/services/haproxy/README.md]: typo available

File: nixos/services/haproxy/config-options.nix [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    default = {};
+    type = submodule {
+      options = defaultsOptions;
+    };
+    description = ''
+      Configuration statements for the defaults section.
+    '';
+  };
+  listens = mkOption {
+    default = {
```
Comment (dpausp) [inline nixos/services/haproxy/config-options.nix]: this shouldn't be a default, move it to `example`

File: nixos/services/haproxy/config-options.nix [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    default = {};
+    type = attrsOf (submodule {
+      options = frontendOptions;
+    });
+    description = ''
+      Frontend sections with statements.
+    '';
+  };
+  backends = mkOption {
+    default = {
```
Comment (dpausp) [inline nixos/services/haproxy/config-options.nix]: this shouldn't be a default, move it to `example`

File: nixos/roles/graylog.nix (hunk 3 above)
Comment (dpausp) [reply nixos/roles/graylog.nix]: Add comment 

~~~
# Journalbeat uses long-running connections and may send nothing
# for a while. Use ttl 120s for Journalbeat to make sure it
# reconnects before it's thrown out by HAproxy.
~~~

File: nixos/services/haproxy/README.md (hunk 5 above)
Comment (dpausp) [reply nixos/services/haproxy/README.md]: change "these modules", it's unclear

Comment (dpausp) [reply nixos/services/haproxy/README.md]: refer to config-options.nix where the options are declared

File: nixos/services/haproxy/default.nix [hunk 8]
```diff
@@ -22,13 +112,13 @@ let
...
-  haproxyCfgContent =
+  haproxyCfgContent = (
     if importedCfgContent != modifiedCfgContent
     then lib.info
       ("HAProxy: you can remove the 'stats socket' line from your config."
       + " It's ignored on NixOS 20.09.")
-      modifiedCfgContent
-    else importedCfgContent;
+    else lib.id
+    ) (generatedConfig + "\n" + modifiedCfgContent);
```
Comment (dpausp) [inline nixos/services/haproxy/default.nix]: add a comment in the generated file that /etc/local/haproxy/*.conf starts here

**Review by dpausp — CHANGES_REQUESTED** [review]

File: doc/src/haproxy.rst [hunk 9]
```diff
@@ -0,0 +1,15 @@
+.. _nixos-haproxy:
```
Comment (ctheune) [inline doc/src/haproxy.rst]: Move this to README.

File: nixos/roles/graylog.nix [hunk 10]
```diff
@@ -249,92 +249,93 @@ in
...
+        shared_beats_graylog_timeout = "121s";
+      in {
+        enable = true;
+        enableStructuredConfig = true;
+        frontends = {
+          gelf-tcp-in = {
+            binds = map (addr: "${addr}:${toString gelfTCPHAPort}") (fclib.listenAddresses "ethsrv");
+            mode = "tcp";
+            options = [ "tcplog" ];
+            timeout = {
```
Comment (ctheune) [inline nixos/roles/graylog.nix]: timeout.client = "10s";

File: nixos/roles/graylog.nix [hunk 11]
```diff
@@ -249,92 +249,93 @@ in
...
+              client = shared_beats_graylog_timeout; # should be equal to server timeout
+            };
+            default_backend = "graylog";
+          };
+        };
+        backends = {
+          gelf_tcp = {
+            mode = "tcp";
+            options = [ "httpchk HEAD /api/system/lbstatus" ];
+            timeout = {
```
Comment (ctheune) [inline nixos/roles/graylog.nix]: see above

File: nixos/roles/kubernetes/frontend.nix [hunk 12]
```diff
@@ -18,37 +18,36 @@ let
...
+    enable = true;
+    enableStructuredConfig = true;
+    defaults = {
+      mode = "tcp";
+      options = [
+        "tcplog"
+        "dontlognull"
+      ];
+    };
+    listens = {
```
Comment (ctheune) [inline nixos/roles/kubernetes/frontend.nix]: consider renaming this to "listen"

File: nixos/roles/kubernetes/frontend.nix [hunk 13]
```diff
@@ -18,37 +18,36 @@ let
...
-
-    listen stats
-        mode http
-        bind 127.0.0.1:8000
-        stats uri /
-        stats refresh 5s
-        stats admin if LOCALHOST
-
-    resolvers cluster
+  haproxyCfg = {
```
Comment (ctheune) [inline nixos/roles/kubernetes/frontend.nix]: I'd rather not have this sit in a let.

File: nixos/services/haproxy/README.md [hunk 14]
```diff
@@ -0,0 +1,209 @@
+The haproxy service module allow you to generate a haproxy configuration from structured nix values.
+
+You have several available options under `flyingcircus.services.haproxy` which are defined in `config-options.nix`:
+
+```
+enable
+enableLocalPlainConfig (default false)
```
Comment (ctheune) [inline nixos/services/haproxy/README.md]: defaults are incorrectly documented

File: nixos/services/haproxy/default.nix [hunk 15]
```diff
@@ -6,9 +6,99 @@ let
   cfg = config.flyingcircus.services.haproxy;
   fclib = config.fclib;
 
+  # last returns the last element of a list or null if the list is empty
```
Comment (ctheune) [inline nixos/services/haproxy/default.nix]: please move the generic functions to nixos/lib in our overlay and access them through fclib.

**Review by ctheune — CHANGES_REQUESTED** [review]

File: nixos/services/haproxy/README.md [hunk 16]
```diff
@@ -0,0 +1,220 @@
+The HAproxy Web Proxy is enabled when the `webgateway` role is enabled and can
```
Comment (dpausp) [inline nixos/services/haproxy/README.md:1]: This is a mixture between restructured text, convert to markdown

File: nixos/services/haproxy/README.md [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+also be enabled through a custom Nix Configuration Option (see :ref:`nixos-custom-modules`).
+
+If Haproxy is enabled it can be configured by adding
+:file:`/etc/local/haproxy/*.conf` files in the HAproxy format as well as
+from custom Nix Configuration Options.
+
+The Nix Configuration Options are documented at
+:file:`./nixos/services/haproxy/README.md` in the `fc-nixos` Platform
```
Comment (dpausp) [inline nixos/services/haproxy/README.md]: references itself

File: nixos/services/haproxy/README.md [hunk 18]
```diff
@@ -0,0 +1,220 @@
+The HAproxy Web Proxy is enabled when the `webgateway` role is enabled and can
+also be enabled through a custom Nix Configuration Option (see :ref:`nixos-custom-modules`).
+
+If Haproxy is enabled it can be configured by adding
+:file:`/etc/local/haproxy/*.conf` files in the HAproxy format as well as
+from custom Nix Configuration Options.
```
Comment (dpausp) [inline nixos/services/haproxy/README.md:6]: explain how to enable structured config or disable plain config, what's the default, and what happens if both are enabled

File: nixos/services/haproxy/frontend.nix [hunk 19]
```diff
@@ -0,0 +1,132 @@
+# This role allows access to Kubernetes Service IPs (also called ClusterIPs, 10.0.0.0/24)
```
Comment (dpausp) [inline nixos/services/haproxy/frontend.nix]: this file shouldn't be here, belongs to the kubernetes role. Just remove it?

**Review by dpausp — CHANGES_REQUESTED** [review]

File: nixos/services/haproxy/frontend.nix (hunk 19 above)
Comment (elikoga) [reply nixos/services/haproxy/frontend.nix]: Whoops, VSCode ist interressant

**Comment by elikoga** [comment]
Tests failing right now would need to debug

**Comment by dpausp** [comment]
merged with small modifications and resolved conflicts in PR #318

**Pull request closed without merging**